[FIX][Core][Checkout] Adding the isEnabled filter to the countries list in checkout while addressing.

christopherhero opened the pull request:
This change restricts the list of available countries in the checkout->addressing step to those which are enabled. 
Currently, when completing the address in checkout, all countries that are assigned to a channel are displayed, regardless of whether the country is enabled (country.enabled). 

| Q                         | A                                    |
|-----------------|-----------------------|
| Branch?              | 2.0 & 1.14                   |
| Bug fix?              | yes                                |
| New feature?     | no                                  |
| BC breaks?        | no                                  |
| Deprecations?   | no                                  |
| Related tickets  | X                                    |
| License              | MIT                                |




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Modified the country selection in address forms to display only active, enabled countries.
	- Enhanced channel functionality to accurately retrieve enabled countries for better user experience.
- **Tests**
	- Added tests to ensure that only enabled countries appear, verifying the new selection logic.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

vvasiloi commented:
Hey @christopherhero, could you please also cover this: https://github.com/Sylius/Sylius/blob/baa1311a89b9f11c5c80a589222f278908064d45/src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php#L53-L57

vvasiloi commented on src/Sylius/Component/Core/Model/Channel.php:
```diff
@@ -295,4 +295,9 @@ public function setMenuTaxon(?TaxonInterface $menuTaxon): void
+
+    public function getEnabledCountries(): Collection
+    {
+        return $this->getCountries()->filter(fn (CountryInterface $country): bool => true === $country->isEnabled());
```
How about using criteria matching?
```suggestion
        return $this->getCountries()->matching(new Criteria(Criteria::expr()->eq('enabled', true)));
```

vvasiloi replied:
Right now the channel countries have eager fetch configured, but it might change in https://github.com/Sylius/Sylius/pull/14286

diimpp replied:
> How about using criteria matching?

Following simpler is better paradigm, filter might be more readable and more common occurrence in the source codes, there is only single `matching()` in the whole codebase.

```php
return $zoneMembers->matching(Criteria::create()->where(Criteria::expr()->eq('code', $code)))->first();
```
> Right now the channel countries have eager fetch configured, but it might change in https://github.com/Sylius/Sylius/pull/14286

Should it be replaced with `$countryRepository->findEnabledForChannel` instead?

christopherhero replied:
> Should it be replaced with $countryRepository->findEnabledForChannel instead?

What you mean ?

christopherhero replied:
Thanks for PR @vvasiloi ! 

For me the `filter` is more consistent with the rest of the code.

diimpp replied:
@christopherhero in case of bigger collection relations, it's more optimal to make repository call instead of processing collection directly, but that's more of theoretical question what should be done for EAGER removal, if it will happen.

diimpp commented:
I've reproduced the issue, it can be re-tagged as bug. /cc @jakubtobiasz 
1. Select few countries for a channel
![image](https://github.com/Sylius/Sylius/assets/870747/05cf957c-ad64-401f-b91e-a8f5b2ff0f60)
2. Start checkout addressing step and see only previously selected countries as available countries list as expected.
![image](https://github.com/Sylius/Sylius/assets/870747/53ac8c53-3609-4afe-8444-e0f3355bfae7)
3. Disable one of the countries, repeat checkout addressing step, it continues to show both enabled and disabled countries list instead of showing only enabled countries.
![image](https://github.com/Sylius/Sylius/assets/870747/4db16cbe-8338-494d-afd1-d5df5a509a98)
![image](https://github.com/Sylius/Sylius/assets/870747/65be07d9-93a5-42f5-aaac-d85d184a9973)

diimpp commented on src/Sylius/Bundle/CoreBundle/Form/Extension/AddressTypeExtension.php:
```diff
@@ -43,7 +43,7 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
             $oldCountryCodeField->getName(),
             $oldCountryCodeField->getType()->getInnerType()::class,
-            array_replace($oldCountryCodeField->getOptions(), ['choices' => $channel->getCountries()->toArray()]),
+            array_replace($oldCountryCodeField->getOptions(), ['choices' => $channel->getEnabledCountries()->toArray()]),
```
There is if clause above this snippet, which most likely should be updated as well.

```patch
---        if ($channel === null || $channel->getCountries()->count() === 0) {
+++        if ($channel === null || $channel->getEnabledCountries()->count() === 0) {
```

christopherhero commented:
> Hey @christopherhero, could you please also cover this:
https://github.com/Sylius/Sylius/blob/baa1311a89b9f11c5c80a589222f278908064d45/src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php#L53-L57

@vvasiloi I will prepare another PR with the fix for the API.
In a few more places we could have used `getEnabledCountries` instead of `getCountries` but in this PR I focused solely on fixing [this case described by](https://github.com/Sylius/Sylius/pull/15100#issuecomment-1716307406) @diimpp

diimpp commented:
PR looks good code-wise, but I wonder should there be a scenario to cover those cases?

On separate note,
- Administrator might wonder why channel has a country, but it's not showing up in checkout
- What about address book, if previously saved country is no longer available, but address is applied to checkout? (Does address book respect countries list per channel at all?)
- Should address book page hide address for disabled country or indicate unavailability somehow else?
- Is there a validation for applied countries in checkout?
- How zones treat disabled countries?
- Already mentioned api checkout

It sounds like too many places needs to be modified to handle disabled counties within relation.
As an idea, what about disallowing disabling countries, if country is in use? It's a simpler coding solution, but not a great administrator experience, when country might be in use by tens of zones and several channels, one might want to exclude country with single click. (And still will require a solution for address book)

diimpp commented on src/Sylius/Component/Core/Model/ChannelInterface.php:
```diff
@@ -89,4 +89,11 @@ public function addCountry(CountryInterface $country): void;
     public function hasCountry(CountryInterface $country): bool;
+
+    /**
+     * @return Collection|CountryInterface[]
+     *
+     * @psalm-return Collection<array-key, CountryInterface>
```
Since psalm was dropped from the project
```suggestion
     * @return Collection<array-key, CountryInterface>
```

christopherhero committed:
[FIX][ChannelBundle][CoreBundle] Adding the isEnabled filter to the list of countries in the addressing step on the cart
- add getEnabledCountries to channel core model
- use channel.getEnabledCountries instead of getCountries in Sylius\Bundle\CoreBundle\Form\Extension\AddressTypeExtension

christopherhero committed:
[FIX][ChannelBundle][CoreBundle] Adding the isEnabled filter to the list of countries in the addressing step on the cart
- add test case to cover "getEnabledCountries" method in Channel core model

TheMilek commented:
Hi!

I'm closing this PR, as it was implemented in https://github.com/Sylius/Sylius/pull/18317, thank you very much for your effort ! 🏅
